### DISCUSSION TITLE
How to use my custom EA with this tool?

### DISCUSSION THREAD

**Discussion by NirobNabil** [EA31337/EA-Tester#230] [Q&A]
I have a custom EA written in python in compiled format for mt5. I need to backtest this EA on large amount of data and get back the results. Any idea how i should proceed?
Thank you

**Comment by kenorb** [comment]
If you've your EA in .ex4 format, then you can use this project.
Check README for syntax and available options.
Here is some demo: https://asciinema.org/a/233268
To get the results, you can export them into the JSON format.
You can also check [MQL-Tester-Action](https://github.com/FX31337/MQL-Tester-Action) project which integrates this EA tester into GitHub Action. So using your self-hosted runners, you can scale up with your testing.
Let me know what you've tried so far.

**Reply by NirobNabil** [reply]
Hello, thanks for the reply. I am looking into the projects you suggested and will let you know soon how it goes. 
And by the way my EAs are in .ex5 format. will these work with your tool?

**Reply by kenorb** [reply]
The .ex5/MT5 formats are not supported, but it can be added with a bit of changes to make it support for MT5 platform (there are already changes for MT5, including config files, but it was never tested).
For MT5 support, I'm working on new tool ([`ansible-role-mt-runner`](https://github.com/EA31337/ansible-role-mt-runner) project) which aims to run .ex5 EAs, but it's not ready yet (there are still few challenges to solve). This needs to run using Ansible.

**Reply by NirobNabil** [reply]
Hello, thats sounds great. I was trying to build the docker locally but faced an error from line 11 of scripts/provisions.sh. I can also contribute to the repository to add mt5 support. Can you perhaps briefly guide me through the proces? Or i can help with the ansible based project too if that can enable me to run backtests on mt5 EAs on cloud platforms.

**Reply by kenorb** [reply]
EA-Tester with all its shell code become too complex, so for MT5 I would recommend Ansible solution, it's fairly simple to use (especially if you want to use cloud), but needs some extra fixes and implementation. You install Ansible, clone the role repo, and play it (run it). More details I've posted at https://github.com/EA31337/ansible-role-mt-runner/discussions/3 (with a Colab demo). If you get stuck at some specific role with some errors, raise the issue at the given repo. Currently I've got some issues installing MT5, needs some debugging.
If you choose EA-Tester solution, then have a look at `install_mt5.sh` (this installs MT5). Once you run `run_backtest` with `-M 5.0.0.3815` (as example), it should download and install platform (not sure if that works), then backtesting script will populate `conf/mt5-tester.ini` with values and start the platform for the test.

**Reply by NirobNabil** [reply]
Great. I'm currently learning Ansible to get along with the new repo.
But for now can guide me through how i can run a locally stored EA through EA-tester with my custom backtest data? I tried giving the absolute directory of my EA through the -e flag when running the docker like the examples but it says it cannot find the EA file. But it works fine with the github link.

**Reply by kenorb** [reply]
When you use `-e`, it's looking for the file within the Docker container (not on your local directory). To access file outside of your container, you've at least 2 ways: use URL or use `-v` to mount your local dir into container (e.g. `docker run -v "$PWD":/opt/EA`). For the example, you can check this [`docker-compose.yml`](https://github.com/EA31337/EA31337/blob/v2.013.1/docker/backtest/Advanced/all-yearly/2021/docker-compose.yml) (which already includes mounting the local directory).

**Reply by NirobNabil** [reply]
I downloaded a dummy mt4 EA from internet and uploaded the EA to a temporary file hosting service and ran the command 
`docker run ea31337/ea-tester:EURUSD-2018-DS run_backtest -v -e https://file.io/ftw2luNl71xq`

[This](https://www.forexstrategiesresources.com/app/download/5375986463/combo242.zip?t=1556393533) is the EA i downloaded 

It gave the following error. Any idea what i am doing wrong? 

<details>

```
Loading /opt/scripts/.aliases.inc.sh... 
Loading /opt/scripts/.funcs.inc.sh...
Loading /opt/scripts/.funcs.cmds.inc.sh... 
Loading /opt/scripts/.funcs.sets.inc.sh...
INFO: Checking platform...
Copying ini files...
INFO: Installed Terminal: 4.0.0.1260
INFO: Installed MetaEditor: 5.0.0.2302
'/opt/conf/mt4-tester.ini' -> '/home/ubuntu/.wine/drive_c/MetaTrader 4/config/mt4-tester.ini'
'/opt/conf/terminal.ini' -> '/home/ubuntu/.wine/drive_c/MetaTrader 4/config/terminal.ini'
INFO: Locating TestExpert file (https://file.io/ftw2luNl71xq)...
--2023-09-30 17:40:38--  https://file.io/ftw2luNl71xq
Resolving file.io (file.io)... 18.205.99.243, 54.86.166.179
Connecting to file.io (file.io)|18.205.99.243|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/octet-stream]
Saving to: '/home/ubuntu/.wine/drive_c/MetaTrader 4/MQL4/Experts/ftw2luNl71xq'

     0K .......... .......... .......... .......... .......    84.5K=0.6s

2023-09-30 17:40:40 (84.5 KB/s) - '/home/ubuntu/.wine/drive_c/MetaTrader 4/MQL4/Experts/ftw2luNl71xq' saved [48264]

INFO: TestExpert file found at: ftw2luNl71xq
INFO: Setting '^TestExpert' to 'ftw2luNl71xq' in mt4-tester.ini
INFO: Configuring start test period (2018.1.01)...
INFO: Setting '^TestFromDate' to '2018.1.01' in mt4-tester.ini
INFO: Configuring end test period (2018.12.31)...
INFO: Setting '^TestToDate' to '2018.12.31' in mt4-tester.ini
INFO: Configuring symbol pair (EURUSD)...
INFO: Setting '^TestSymbol' to 'EURUSD' in mt4-tester.ini
Getting 'TestModel' from mt4-tester.ini: 0
Getting 'TestExpert' from mt4-tester.ini: ftw2luNl71xq
Getting 'Expert' from mt4-tester.ini:
Getting 'Script' from mt4-tester.ini: 
Getting 'Server' from mt4-tester.ini:
'/opt/conf/ea.ini' -> '/home/ubuntu/.wine/drive_c/MetaTrader 4/tester/ftw2luNl71xq.ini'
Getting 'Server' from mt4-tester.ini:
'/home/ubuntu/.wine/drive_c/MetaTrader 4/history/default/symbols.raw' -> '/home/ubuntu/.wine/drive_c/MetaTrader 4/history//symbols.raw'
INFO: Configuring test report (tester/ftw2luNl71xq-Report.htm)...
INFO: Setting '^TestReport' to 'tester/ftw2luNl71xq-Report.htm' in mt4-tester.ini
INFO: Checking destination directory (/opt/results)...
Getting 'TestPeriod' from mt4-tester.ini: M30
INFO: Checking backtest data (DS)...
Getting 'bt_data' from custom.ini: EURUSD-2018-DS
_XSERVTransmkdir: Owner of /tmp/.X11-unix should be set to root

The VNC desktop is:      4fb64aedd80b:0
PORT=5900
INFO: IP: 172.17.0.2  (4fb64aedd80b)
INFO: Cleaning previous test data...
INFO: Starting...
...
0       17:40:58.059    Started with configuration file 'config/mt4-tester.ini'
INFO: Checking logs for warnings and errors...
tester/logs/20230930.log:3      17:41:03.561    cannot open file 'C:\MetaTrader 4\MQL4\Experts\ftw2luNl71xq.ex4' [2]
tester/logs/20230930.log:3      17:41:03.561    Tester: cannot start with configuration file because expert ftw2luNl71xq not loaded
ERROR: RUN failed with errors.
1       17:41:05.870    '12345': connect failed [No connection]
0       17:41:06.818    MetaTrader 4 build 1260 stopped
0       17:40:51.601    Compiling 'TestSymgroupsRaw'
0       17:40:51.601    Compiling 'TestTicksRaw'
0       17:40:51.601    Compiling 'TestTimeframes'
0       17:40:51.602    Compiling 'DLLSampleTester'
0       17:40:51.604    Compiling 'PipeClient'
0       17:40:51.604    Compiling 'PipeClientPure'
0       17:40:51.605    Compiling 'Dummy'
0       17:40:51.605    Compiling 'PrintPaths'
0       17:40:51.605    Compiling 'TestInitFail'
0       17:40:56.193    Compiling 'stdlib'
Xvfb killed (pid 174)
ERROR: Exiting /opt/scripts/run_backtest.sh with 1
415 on_error /opt/scripts/.funcs.inc.sh
34 on_success /opt/scripts/run_backtest.sh
888 main /opt/scripts/run_backtest.sh
XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":0"
      after 193 requests (193 known processed) with 0 events remaining.
XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":0"
      after 12 requests (11 known processed) with 0 events remaining.
Exiting /opt/scripts/run_backtest.sh with 1
```

</details>

**Reply by kenorb** [reply]
file.io site is not ideal, because it takes you to the download page without downloading the file it-self and it needs manual user intervention. Try uploading .ex4/.ex5 into GitHub, but they won't support .ex4/ex5 extensions in issues or discussion. Check code of `dl_ea.sh` script which is doing the download.
Actually the logic for downloading EA from URL you can find in `.funcs.cmds.inc.sh` (check `ea_find` function). If you want to debug scripts, add `-x` param.

**Reply by kenorb** [reply]
This EA-Tester project has been now destroyed by MQ legal attacks targeting and destabilizing our infrastructure.
Read more: https://github.com/github/dmca/blob/master/2023/11/2023-11-07-metatrader.md
Not sure if this project can recover from it, we'll see.
So the suggested way of backtesting is to use Ansible method now.

**Reply by NirobNabil** [reply]
Thanks for letting me know. I've already switched to the ansible platform. It has been working great so far.
